backend/api/controllers/OrderController.js

```diff
@@ -0,0 +1,143 @@
+    User.findOne(req.session.userId).exec(function (err, user) {
+      if (err) return res.negotiate(err);
+      if (!user) return res.unauthorized();
+
+      // Fetch product details from DB
+      // TODO don't allow buying unlisted products
+      Product.find({id: Object.keys(orderQuantity)}).exec(function (err, products) {
+        if (err) return res.negotiate(err);
+        if (!products) return res.badRequest({error: 'None of the products you ordered exist (we think)'});
+        if (products.length !== req.body.length) return res.badRequest({error: (req.body.length - products.length) +
+          ' product(s) in your request do not exist or your request contains dupes'});
+
+        // Accumulate total price for order
+        let priceAccumulator = 0;
+        // Details about this order that is returned to the user
+        let orderDetails = {};
+        orderDetails.products = [];
+
+        // Iterate over our products and calculate price
+        for (let product of products) {
+
+          let quantity = orderQuantity[product.id];
+          let productPrice = 0;
+
+          switch (product.on_sale) {
+            case 'PRICE_MOD': // Percent sale
+              productPrice = quantity * product.price * product.price_mod;
+              break;
+
+            case 'PACKAGE': // a for b package deal
+              // Calculate total price for number of full packages and the remaining number
+              let packageCount = Math.floor(quantity / product.package_get_count);
+              let remainder = quantity % product.package_get_count;
+              // Price to pay for full packages
+              productPrice = packageCount * product.package_pay_count * product.price;
+              // Add price for remaining product quantity
+              productPrice += remainder * product.price;
+              break;
+
+            default: // Full price, just for you
+              productPrice = quantity * product.price;
+          } // switch
+
+          // Add product price sum rounded to two decimals
+          productPrice = +productPrice.toFixed(2);
+          priceAccumulator += productPrice;
+
+          orderDetails.products.push({product: product, quantity: quantity, sum: productPrice});
+        } // for products end
+
+        orderDetails.totalPrice = priceAccumulator;
+        orderDetails.orderNumber = null;
+
+        // Create a new order
+        Order.create({user: user.id, total_price: priceAccumulator, order_details: orderDetails}).exec(
+          function (err, order) {
+          if (err) return res.negotiate(err);
+          if (!order) return res.serverError();
+
+          order.order_details.orderNumber = order.id;
+          order.order_details.orderDate = order.createdAt;
+          order.save(function (err) {
+            return res.created(order.order_details);
+          });
+
+        }); // Order callback end
+      }); // Product callback end
+    }); // User callback end
+
```

Comment:
#### General feedback.
In order to be maintainable, this function `create` needs some improvements. The function  is too large. With 88 lines, it is so long that the developer found it necessary to comment each block ending. That is a bad sign. 

#### Use Promises instead of callbacks
In this functions, there are three callbacks inside of each other. That can quickly create a mess - and all of your code disappears out of the screen to the right, because of all the indentation. This problem is largely solved if you make use of Promises.

#### Use `forEach`, `map`, `filter` and `reduce` instead of iterators
Javascript has excellent built-in functions for dealing with arrays - amongst them are `forEach`, `map`, `filter` and `reduce`. Learning those functions and when to use them would create more readable and maintainable code.

#### Split the function into several smaller functions
Every function should ideally fit into one computer screen, and should only do *one* thing. This function could be split into several smaller functions. One for finding the user. One for finding the products. One for calculating the price of a product based on sales. And one for creating an order. Then the next developer who sees this code would easier understand what is going on here.

#### Use more brackets
Many if statements in this functions does not use brackets. While somewhat frowned upon, dropping the brackets in small one-liners is ok - but it is important that the entire statement is on one line. Otherwise it can be easy to create confusing code, like this:

```javascript
if (x > 3)
    x = 5;
    y = 2;
```